lib/set.rb

```diff
@@ -361,16 +361,15 @@ def to_set(klass = Set, *args, &block)
     klass.new(self, *args, &block)
   end
 
-  def flatten_merge(set, seen = Set.new) # :nodoc:
+  def flatten_merge(set, seen = Set.new.compare_by_identity) # :nodoc:
     set.each { |e|
       if e.is_a?(Set)
-        if seen.include?(e_id = e.object_id)
+        unless seen.add?(e)
```

Comment:
There's 3 nice performance wins here:

* We don't trigger a look up (and potentially an assignment) of an object ID
* We can do an identity-based look-up, which is faster
* We combine the `include?` and `add` steps into one, saving one hash lookup/comparison.